contents/code/tilingmanager.js

```diff
@@ -975,3 +1013,80 @@ TilingManager.prototype._getLayouts = function(desktop, screen) {
     }
     return null;
 };
+
+TilingManager.prototype._isDesktopEmpty = function(desktop) {
+    var clients = workspace.clientList();
+    var empty = true;
+    for (var i = 0; i < clients.length; i++) {
+        if (!clients[i].onAllDesktops && clients[i].desktop == desktop) {
+            empty = false;
+            break;
+        }
+    }
+    return empty;
+};
+
+TilingManager.prototype._compactDesktops = function() {
+    this._compacting = true;
+    for (var destination = 1; destination < this.desktopCount; destination++) {
```

Comment:
Okay, so this is the heart of the compacting idea.

It walks all desktops, and for each empty desktop, it removes it and moves all later desktops one back.

So if your desktops look like ("F" for full, "E" for empty):

FEFEFEF

it moves:

```
F(FEFEF)
FF(FEF)
FFF(F)
```

all those desktops in parentheses.

Would it work better to walk this backwards?

That would mean

```
FEFEFEF
FEFEF(F)
FEF(FF)
F(FFF)
```

i.e.  a total of 6 moves instead of 9, because we're not moving the empty desktops multiple times.

Either that, or move through it once, make a new list with non-empty desktops, move through that and fix the number.

Reply:
>Would it work better to walk this backwards?

Nope. I didn't want the desktop order to get altered. The back-to-front algorithm would have to be recursive in order to shift the populated desktops instead of swapping them.

>Either that, or move through it once, make a new list with non-empty desktops, move through that and fix the number.

Yes, could be done. I could have constructed an {old: new} map of desktop indices and move each only once afterwards.

Reply:
> Nope. I didn't want the desktop order to get altered. The back-to-front algorithm would have to be recursive in order to shift the populated desktops instead of swapping them.

Ah, sorry, I missed a bit there! I thought you were moving _all_ following desktops, but you're just swapping with the next non-empty one, so all the empty desktops are at the end.

Which makes much more sense!

Reply:
Yes, thinking about it now, is should not happen to move one desktop several times.